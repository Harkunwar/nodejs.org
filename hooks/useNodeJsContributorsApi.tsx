import useSWR from 'swr';
import { NodeJSContributorsApiKey } from '../constants/swr';

const LIMIT_CONTRIBUTORS = 5;
const CONTRIBUTORS_API_URI = `https://api.github.com/repos/nodejs/node/contributors?per_page=${LIMIT_CONTRIBUTORS}`;

export type ParsedLink = {
  url: string;
  page: number;
};

export type ContributorApiResponse = {
  login: string;
  id: number;
  url: string;
  type: string;
  contributions: number;
  node_id: string;
  avatar_url: string;
  gravatar_id: string;
  html_url: string;
  followers_url: string;
  following_url: string;
  gists_url: string;
  starred_url: string;
  subscriptions_url: string;
  organizations_url: string;
  repos_url: string;
  events_url: string;
  received_events_url: string;
  site_admin: boolean;
};

export type Contributor = {
  avatarUri: string;
  profileUri: string;
  login: string;
  contributionsCount: number;
  commitsListUri: string;
};

/**
 * Parses "Link" response header from node/contributors API
 * Returns page values for "next" and "last" API URLs
 * @param linkHeader
 */
export function linkParser(linkHeader: string): Record<string, ParsedLink> {
  const regex = new RegExp(
    `<([^?]+\\?per_page=${LIMIT_CONTRIBUTORS}&[a-z]+=([\\d]+))>;[\\s]*rel="([a-z]+)"`,
    'g'
  );
  let array: RegExpExecArray | null = null;
  const parsedLinks: Record<string, ParsedLink> = {};

  do {
    array = regex.exec(linkHeader);

    if (array) {
      parsedLinks[array[3]] = {
        url: array[1],
        page: parseInt(array[2], 10),
      };
    }
  } while (array !== null);

  return parsedLinks;
}

/**
 * Retrieves max amount of contributors for Node.js main repo.
 * Returns array with random contributor index and max contributors found.
 */
export async function getMaxContributors(): Promise<[number, number]> {
  const response = await fetch(CONTRIBUTORS_API_URI);
  const linksHeaderValue = response.headers.get('Link');

  if (linksHeaderValue) {
    const links = linkParser(linksHeaderValue);

    const randomPage =
      Math.floor(Math.random() * Math.floor(links?.last?.page ?? 1)) + 1;
    return [randomPage, links?.last?.page ?? 1];
  }

  throw new Error('Failed to get amount if max contributors');
}

/**
 * Retrieves a contributor's object by it's index in API
 * @param randomPage
 */
export async function getContributor(randomPage: number): Promise<Contributor> {
  const response = await fetch(`${CONTRIBUTORS_API_URI}&page=${randomPage}`);
  const jsonResponse = (await response.json()) as ContributorApiResponse[];

  const contributorData: Contributor[] = jsonResponse.map(
    ({ avatar_url, login, contributions, html_url }) => ({
      avatarUri: avatar_url,
      login,
      contributionsCount: contributions,
      profileUri: html_url,
      commitsListUri: `https://github.com/nodejs/node/commits?author=${login}`,
    })
  );

  const contributor = contributorData.shift() as Contributor;

  if (window.localStorage) {
    window.localStorage.setItem(
      'contributors',
      JSON.stringify(contributorData)
    );
  }

  return contributor;
}

/**
 * Calls relative APIs and returns random contributor for Node.js main repo.
 * Trying to store cached data in localStorage in order to do less consequent requests
 */
export async function fetchRandomContributor(): Promise<Contributor> {
  let maxContributors: number | null = null;
  let fetchDate: number | null = null;
  let needToRefetch = false;
  let contributors: Contributor[] = [];
  const ONE_MONTH_MS = 2592000000;

  if (window.localStorage) {
    const maxContributorsStored =
      window.localStorage.getItem('max_contributors');
    const fetchDateStored = window.localStorage.getItem('fetch_date');

    contributors = JSON.parse(
      window.localStorage.getItem('contributors') || '[]'
    );

    maxContributors = maxContributorsStored
      ? parseInt(maxContributorsStored, 10)
      : null;
    fetchDate = fetchDateStored ? parseInt(fetchDateStored, 10) : null;
  }

  if (fetchDate && Date.now() - fetchDate >= ONE_MONTH_MS) needToRefetch = true;

  return new Promise((resolve, reject) => {
    if (contributors && contributors.length > 0 && !needToRefetch) {
      const contributor = contributors.shift();
      if (window.localStorage) {
        localStorage.setItem('contributors', JSON.stringify(contributors));
      }
      resolve(contributor as Contributor);
    } else if (maxContributors && !needToRefetch) {
      getContributor(
        Math.floor(Math.random() * Math.floor(maxContributors)) + 1
      )
        .then(contributor => resolve(contributor))
        .catch(error => reject(error));
    } else {
      getMaxContributors()
        .then(([randomPage]) => getContributor(randomPage))
        .then(contributor => {
          if (window.localStorage) {
            window.localStorage.setItem('fetch_date', String(Date.now()));
          }
          resolve(contributor);
        })
        .catch(error => reject(error));
    }
  });
}

export function useNodeJsContributorsApi(
  isVisible: boolean
): Contributor | null | undefined {
  const { data: contributor } = useSWR(
    isVisible ? NodeJSContributorsApiKey : null,
    fetchRandomContributor
  );
  return contributor;
}
