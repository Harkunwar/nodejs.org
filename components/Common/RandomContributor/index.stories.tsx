import { createMock } from 'storybook-addon-module-mock';
import * as hooks from '../../../hooks/useNodeJsContributorsApi';
import RandomContributor from './index';
import type { Meta as MetaObj, StoryObj } from '@storybook/react';

type Story = StoryObj<typeof RandomContributor>;
type Meta = MetaObj<typeof RandomContributor>;

export const Default: Story = {
  parameters: {
    moduleMock: {
      mock: () => {
        const mock = createMock(hooks, 'useNodeJsContributorsApi');
        mock.mockReturnValue({
          profileUri: 'https://github.com/nodejs',
          avatarUri:
            'https://avatars.githubusercontent.com/u/9950313?s=200&v=4',
          login: 'nodejs',
          contributionsCount: 20,
          commitsListUri:
            'https://github.com/nodejs/node/commits?author=nodejs',
        });
      },
    },
  },
};

export default { component: RandomContributor } as Meta;
