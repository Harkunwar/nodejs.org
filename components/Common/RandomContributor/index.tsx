import { type MutableRefObject, useRef } from 'react';
import { FormattedMessage } from 'react-intl';
import Image from 'next/image';
import styles from './index.module.scss';
import AnimatedPlaceholder from '../AnimatedPlaceholder';
import { useOnScreen } from '../../../hooks/useOnScreen';
import { useNodeJsContributorsApi } from '../../../hooks/useNodeJsContributorsApi';

const DefaultImageSize = 75;

const RandomContributor = () => {
  const ref = useRef<HTMLDivElement | null>(null);
  const isVisible = useOnScreen(ref as MutableRefObject<Element>, true);
  const contributor = useNodeJsContributorsApi(isVisible);

  return (
    <div ref={ref} className={styles.randomContributor}>
      {!contributor && isVisible && <AnimatedPlaceholder />}
      {contributor && (
        <>
          <div className={styles.randomContributorAvatar}>
            <a
              href={contributor.profileUri}
              target="_blank"
              rel="nofollow noopener noreferrer"
            >
              <Image
                src={contributor.avatarUri}
                alt="Avatar of a Node.js contributor"
                width={DefaultImageSize}
                height={DefaultImageSize}
              />
            </a>
          </div>
          <a
            href={contributor.profileUri}
            target="_blank"
            rel="nofollow noopener noreferrer"
          >
            <FormattedMessage
              id="components.randomContributor.thankYou"
              values={{
                contributor: contributor.login,
                amount: contributor.contributionsCount,
              }}
            />
          </a>
        </>
      )}
    </div>
  );
};

export default RandomContributor;
