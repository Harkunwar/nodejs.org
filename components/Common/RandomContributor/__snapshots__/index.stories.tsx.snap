// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Common/RandomContributor Default smoke-test 1`] = `
<div>
  <div>
    <a href="https://github.com/nodejs"
       target="_blank"
       rel="nofollow noopener noreferrer"
    >
      <img alt="Avatar of a Node.js contributor"
           loading="lazy"
           width="75"
           height="75"
           decoding="async"
           data-nimg="1"
           srcset="https://avatars.githubusercontent.com/u/9950313?s=75&amp;v=4?w=96&amp;q=75 1x, https://avatars.githubusercontent.com/u/9950313?s=75&amp;v=4?w=256&amp;q=75 2x"
           src="https://avatars.githubusercontent.com/u/9950313?s=75&amp;v=4?w=256&amp;q=75"
           style="color: transparent;"
      >
    </a>
  </div>
  <a href="https://github.com/nodejs"
     target="_blank"
     rel="nofollow noopener noreferrer"
  >
    Thank you nodejs for being a Node.js contributor with 20 contributions
  </a>
</div>
`;

exports[`Common/RandomContributor Loading smoke-test 1`] = `
<div>
</div>
`;
